### PR TITLE
QuickSort3 Implementation

### PR DESCRIPTION
**Pull request by AtharvaThorve**
As stated in issue #183. This is the implementation of the 3 Way QuickSort algorithm or the Dutch National Flag algorithm.

### CODE REVIEW

File: src/main/java/com/williamfiset/algorithms/sorting/QuickSort3.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      if (a[r] < a[l]) {
+        int temp = a[l];
+        a[l] = a[r];
+        a[r] = temp;
+      }
+      j = l;
+      k = r;
+      int m1 = j;
+      int m2 = k;
+      int[] m = {m1, m2};
```
Comment (williamfiset) [inline src/main/java/com/williamfiset/algorithms/sorting/QuickSort3.java:41]: Do you need m1 and m2 as temps here? I _think_ j and k would be copied by value, same below

File: src/main/java/com/williamfiset/algorithms/sorting/QuickSort3.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    QuickSort3.randomizedQuickSort(ar, 0, ar.length - 1);
+  }
+
+  // partiton array in such a way that all the elements whose value is equal to
+  // pivot are grouped together
+  private static int[] partition3(int[] a, int l, int r) {
+    int j, k;
+    if (r - l <= 1) {
+      if (a[r] < a[l]) {
+        int temp = a[l];
```
Comment (williamfiset) [inline src/main/java/com/williamfiset/algorithms/sorting/QuickSort3.java:33]: Maybe a swap helper function would be nice? You're using this swap paradigm in multiple places.

https://github.com/williamfiset/Algorithms/blob/master/src/main/java/com/williamfiset/algorithms/sorting/QuickSort.java#L50

File: src/main/java/com/williamfiset/algorithms/sorting/QuickSort3.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
+/**
+ * Quicksort implementation using Hoare partitioning
```
Comment (williamfiset) [inline src/main/java/com/williamfiset/algorithms/sorting/QuickSort3.java:2]: I think your description in https://github.com/williamfiset/Algorithms/issues/183 is more accurate?

**Comment by williamfiset** [comment]
Thanks, really cool twist on quicksort! I'll add it to the README once checked in

File: src/main/java/com/williamfiset/algorithms/sorting/QuickSort3.java (hunk 1 above)
Comment (AtharvaThorve) [reply src/main/java/com/williamfiset/algorithms/sorting/QuickSort3.java:41]: Yess okay I'll add that change.

File: src/main/java/com/williamfiset/algorithms/sorting/QuickSort3.java (hunk 3 above)
Comment (AtharvaThorve) [reply src/main/java/com/williamfiset/algorithms/sorting/QuickSort3.java:2]: Ohh yeah sorry I'll do it right away.

**Comment by williamfiset** [comment]
Thank you @AtharvaThorve !

**Pull request merged**